File: 2022/day03/day03.jl
function priority(x)
    return islowercase(x) ? x - 'a' + 1 : x - 'A' + 27
end

function part1(data)
    p = 0
    for line in data
        s = length(line) ÷ 2
        a, b = line[1:s], line[end-s+1:end]
        p += priority(only(a ∩ b))
    end
    return p
end

function part2(data)
    p = 0
    for row in eachcol(reshape(data, (3, :)))
        p += priority(only(intersect(row...)))
    end
    return p
end

function main()
    data = filter(x -> !isempty(x), split(readchomp("input.txt"), "\n"))
    @show part1(data)
    @show part2(data)
end

main()

File: 2022/day03/day03.py
data = open("input.txt").read().strip().split()

# part 1
p = 0
for line in data:
    a, b = line[: len(line) // 2], line[-len(line) // 2 :]
    unique = (set(a) & set(b)).pop()
    print(ord("a"), ord("A"))
    print(unique, ord(unique), ord(unique) - ord("A") + 27)
    exit()
    if unique.islower():
        p += ord(unique) - ord("a") + 1
    else:
        p += ord(unique) - ord("A") + 27
print(p)

# part 2
p = 0
for a, b, c in zip(data[::3], data[1::3], data[2::3]):
    badge = (set(a) & set(b) & set(c)).pop()
    if badge.islower():
        p += ord(badge) - ord("a") + 1
    else:
        p += ord(badge) - ord("A") + 27
print(p)
